resources/todomvc/architecture-examples/angular/src/app/todos.service.ts

```diff
@@ -0,0 +1,55 @@
+import { Injectable } from "@angular/core";
+import { Todo } from "./todo";
+
+function uuid() {
+    let uuid = "";
+    for (let i = 0; i < 32; i++) {
+        const random = (Math.random() * 16) | 0;
+        if (i === 8 || i === 12 || i === 16 || i === 20)
+            uuid += "-";
+
+        uuid += (i === 12 ? 4 : i === 16 ? (random & 3) | 8 : random).toString(16);
+    }
+    return uuid;
+}
+
+@Injectable({
+    providedIn: "root",
+})
+export class TodosService {
+    todos: Todo[] = [];
+
+    addItem(title: string): void {
+        const todo: Todo = {
+            id: uuid(),
+            title,
+            completed: false,
+        };
+
+        this.todos = this.todos.concat(todo);
+    }
+
+    removeItem(todo: Todo): void {
+        const index = this.todos.findIndex((t) => t.id === todo.id);
+        this.todos.splice(index, 1);
+    }
+
+    clearCompleted(): void {
+        this.todos = this.todos.filter((todo) => !todo.completed);
+    }
+
+    toggleAll(completed: boolean): void {
+        this.todos = this.todos.map((todo) => ({ ...todo, completed }));
```

Comment:
Can we toggle `completed` by looping instead of creating a new array?
For example we could use `for of`, which would also have the nice side effect of exercizing a different part of the JS engine (the iterable protocol).

Reply:
since the item uses `ChangeDetectionStrategy.OnPush`, this wouldn't work